crates/app-host/src/state_manager.rs

```diff
@@ -0,0 +1,296 @@
+
+use std::{collections::{HashMap, VecDeque}, sync::{Arc, Mutex}};
+
+use base::eth::EthError;
+use rpc::{ProveBatchRequest, ProveBatchResponse, ProveBundleRequest};
+
+use crate::{block_tracer::BlockTracer, l1_client::L1Client, types::{BatchHash, CommitBatchEvent, FinalizeBatchEvent, StateRoot}};
+use anyhow::{bail, Ok, Result};
+use alloy::primitives::Bytes;
+
+base::stack_error! {
+    #[derive(Debug)]
+    name: StateManagerError,
+    stack_name: StateManagerErrorStack,
+    error: {
+        Eth(EthError),
+        Custom(std::borrow::Cow<'static, str>),
+    },
+    wrap: {
+    },
+    stack: {}
+}
+
+impl From<EthError> for StateManagerError {
+    fn from(value: EthError) -> Self {
+        Self::Eth(value)
+    }
+}
+
+struct BatchInfo {
+    batch_index: u64,
+    batch_header: Option<Bytes>,
+    prove_response: Option<ProveBatchResponse>,
+    start_block_number: u64,
+}
+
+struct BatchState {
+    hash_info_map: HashMap<BatchHash, BatchInfo>,
+    index_hash_map: HashMap<u64, BatchHash>,
+}
+
+impl BatchState {
+    fn create_batch(&mut self, event: &CommitBatchEvent, batch_info: BatchInfo) {
+        self.index_hash_map.insert(event.batch_index, event.batch_hash);
+        self.hash_info_map.insert(event.batch_hash, batch_info);
+    }
+
+    fn update_batch_header(&mut self, batch_index: u64, batch_header: Bytes) {
+        if let Some (batch_hash) = self.index_hash_map.get(&batch_index) {
+            self.hash_info_map.entry(*batch_hash).and_modify(|info| info.batch_header = Some(batch_header));
+        };
+    }
+
+    fn update_batch_proof(&mut self, prove_response: ProveBatchResponse) {
+        self.hash_info_map.entry(prove_response.batch_hash).and_modify(|info| info.prove_response = Some(prove_response));
+    }
+
+    // this method requires that the batch should be proved sequentially by enclave part
+    // or it fails to get the prev_state_root
+    fn get_batch_prev_state_root(&self, batch_index: u64) -> Option<StateRoot> {
+        let prev_batch_index = batch_index - 1;
+        self.index_hash_map.get(&prev_batch_index).and_then(|batch_hash| {
+            self.hash_info_map[batch_hash].prove_response.as_ref().and_then(|response| {
+                Some(response.post_state_root)
+            })
+        })
+    }
+
+    fn collect_batch_infos(&self, begin_batch_index: u64, end_batch_index: u64) -> Result<ProveBundleRequest> {
```

Comment:
And how do we construct `ProveBatchRequest`?

Reply:
it's constructed within `on_batch_commit_event_received`